### ISSUE TITLE
more fuzz tests

### ISSUE THREAD

**Issue by pca006132** [elalish/manifold#667]
We only have tests for the triangulator, and the test is only checking for segfault/memory corruption and infinite loops. We should add more tests.

1. [ ] Add tests for the triangulator with valid polygons.
2. [ ] Add tests for CSG.

CSG can be fuzzed with a simple grammar.

**Comment by pca006132** [comment]
probably good dataset: https://ten-thousand-models.appspot.com/results.html?q=is+manifold%2C+is+oriented%2C+with+degenerate

**Comment by dasoya** [comment]
Dear Emmett Lalish,

Thank you for considering my proposal.

I am passionate about working in the entertainment industry, such as
digital fashion using computer graphics. During my internship with NCsoft's
Digital Human team, I utilized C++ in Unreal Engine and TensorFlow to
develop 3D modeling AI. I developed games and had the opportunity to be a
PM and lead the release. My strength is that I take responsibility and get
things done. My technical skills include implementing ray tracing using C++
based on OpenGL. I have a strong interest in the technology that Manifold
implements and am willing to make improvements where needed.

Thank you for your consideration,

Kind regards
Daso Jung.


2024년 4월 17일 (수) 오전 2:13, Emmett Lalish ***@***.***>님이 작성:

**Comment by elalish** [comment]
@dasoya That's wonderful, and we'd certainly appreciate your contributions! From the GSoC side, they're much more willing to accept those who have demonstrated successful contribution to the code base. If you'd be willing to open a simple PR we can review and merge, it'll go a long way. If you're not sure where to start, a simple task we've been meaning to get around to is splitting our tests into more files - there are so many now that the granularity is much too coarse. If you'd like to take a stab at organizing some of them a bit, we'd certainly appreciate it.

**Comment by dasoya** [comment]
Thank you for your suggestion. I’ll try to contribute to the code base!

Daso Jung

**Comment by dasoya** [comment]
@elalish  As you suggested, I thought boolean_test.cpp has high complexity, so I divided it into two files.
I'm going to make additional boolean tests in the complex_boolen_test.cpp or new fuzz test file.

https://github.com/elalish/manifold/compare/master...dasoya:manifold:boolean_fuzz

Is this the right way? Let me know what you think.

**Comment by elalish** [comment]
Yes, that's a great start - go ahead and open a PR. It's always best to have many, smaller PRs anyway, and it gives us a good place to discuss.

**Comment by zmerlynn** [comment]
Hi - as part of #1707 I put together a FuzzTest harness for the 2D side (libFuzzer compat mode, per-target corpus that carries forward across CI runs). Before upstreaming it or extending it to 3D CSG, wanted to sanity-check the operational side.

The harness lives on the fork: [test/cross_section_fuzz.cpp](https://github.com/zmerlynn/manifold/blob/pr/cross-section-fuzz/test/cross_section_fuzz.cpp) with a [CI workflow](https://github.com/zmerlynn/manifold/blob/local/fork-ci-fuzz/.github/workflows/fuzz_continuous.yml) that runs each target independently with corpus caching. I've been running a producer/consumer campaign on top of it -- one session finds bugs and drops them as `DISABLED_` regression seeds, another drains and fixes them. It's been effective for the 2D work.

The main difference from the existing fuzzers: rather than crash/hang detection, it asserts algebraic properties (commutativity, inclusion-exclusion, associativity, scale/rotation/translation invariance, etc.). That's caught real bugs that wouldn't have surfaced as crashes. The same approach should extend pretty naturally to `manifold_fuzz.cpp` on top of `SimpleCube`.

A continuous campaign tends to produce a steady drip of findings -- some real, some noise -- and without a clear owner and triage path it piles up fast. What's the preferred way to file findings if something surfaces? Open issues, draft PRs, something else? And for 3D: is there appetite for a sustained campaign, or more of a one-shot "run until clean"?

**Comment by elalish** [comment]
This looks closely related to @AnshulPatil2005's work, so it'd be good to get yours and @pca006132's opinions here. If the fuzz produces a nice `TEST` case automatically, then a PR adding that would be ideal. If it's non-trivial to fix, we can always merge it with `DISABLED_` to act as a TODO.

**Comment by pca006132** [comment]
But do we really have things like commutativity and associativity? Do you expect exactly the same output or just test using other properties, e.g., genus, surface area, volume?

I haven't yet thought too much about a sustained campaign. That said, with @AnshulPatil2005's work, maybe we can do a weekly run, which is quite useful when we have more weekly changes and don't have to rely on maintainers to run them locally (which we should do more often...).
